### ISSUE TITLE
Processing of tentatives/unknowns

### ISSUE THREAD

**Issue by sneumann** [MassBank/RMassBank#42]
_Issue by **[schymane](https://github.com/schymane)** from Monday Oct 28, 2013 at 13:42 GMT_
_Originally opened as https://github.com/sneumann/RMassBank/issues/34_

---

This relies on other discussions to set the tagging. TBC.

**Comment by schymane** [comment]
Talk about this after BioC release

**Comment by schymane** [comment]
Erik and I also discussed this today, a summary:
Michele and I both feel that MetFrag *in* RMassBank is out of the scope, certainly for the moment. More fundamental things that need to be addressed first to enable RMassBank to even create tentative/unknown spectra suitable for upload to MassBank. Assuming the *user* knows why they want the spectra and how they identified it we need the right comment fields, the correct way to represent the results, what do we do without structures, how to skip the recalibration and formula annotation without a formula available ... (but still do this if we do have a formula but not structure).
Case 1: we have a structure but it's not certain => this is only a comment in the MassBank record
Case 2: we aren't exactly sure about the structure but we have a formula => can still perform most aspects of the workflow
Case 3: we have an exact mass with a cool spectrum that we want to share with the world, but we can't identify it
Strategy: RMassBank works off a compound list where the user gives SMILES (best case) or for tentatives/unknowns either a formula or mass along with a comment field and RMassBank takes this information and prepares the records as best it can (note, we probably need to deal with mixed lists...)
Once we have the records there is absolutely nothing stopping anyone doing MetFrag on them ... we actually almost have all the bits and pieces ready but this is a different connection/scope and in my opinion the job of another package and not RMassBank at the moment.

**Comment by ermueller** [comment]
So, this is theoretically finished, but it needs cleanup and stuff like that. Docs are already in there, but this is all subject to change, since I already typed up a list for myself on where to clean up some code.

You'll all get a mail today with everything, I'll just test this a bit more and then type up some instructions.

Also, add a vignette for tentatives/unknowns? I feel that we have to. But I'd make that a separate issue.

**Comment by schymane** [comment]
Great!
I agree that we need a vignette and that it should be a separate issue. Look forward to testing it ;)

**Comment by ermueller** [comment]
I hope you all read this, since my uni mailserver half-crashed 5 minutes ago:

I forgot the attachment in my mail, but you can install the package from git from my master branch.
https://github.com/ermueller/RMassBank

**Comment by schymane** [comment]
Minor feedback to your email point 3: @ermueller 
a) If you want to process unknowns you need an additional column "m/z" with your target m/z.
=> Michele and I would be very happy if you take "mz" instead of "m/z" to avoid ugly conversion issues downstream.

**Comment by ermueller** [comment]
No problem, I'll rework that. Can both be valid? :)

Also, I'll probably autodetect what you can do with a certain compound list and give the user an output when the list is loaded. Something like
"This compoundlist can be used for the following retrieval methods: standard, tentative"

Also, I'll save that somewhere internally and make workflow retrieval methods unable if the compoundlist loaded is not appropriate. Sounds good?

**Comment by schymane** [comment]
I’m just trying to wrap my head around your way of building it, because I envisaged doing it all together.
Two conceptually challenging bits:
We’ll have to transfer the recalibration to the tentatives and unknowns from the standard workflow
We have the problem that we now have several runs for one dataset and this is a teeny tiny problem at the record generation stage because the list file is overwritten, see #145.
I’m out of time for now but can get you more feedback tonight or tomorrow.

**Comment by ermueller** [comment]
1) Transfer recalibration for unknowns (without annotation)? How would you calculate (and adjust) dppm for peaks if there are no formulas? Recalibration should work for tentatives.

2) Several runs for one dataset? I'm not sure I understand. Different in what respect? I can def program a multiprocessing in for pretty much every option, as long as the resulting records would have a different accession.

**Comment by schymane** [comment]
Yeah so I was envisaging that you read out from the compound list what the case is and just do what you can (smiles = normal workflow, no smiles but formula = tentative, no smiles+no formula= unknown. No smiles, no formula, no mz = error). Because I have these all mixed together in my one compound list and each one has a unique ID, so there are no accession clashes.
Whether they are actually processed one after the other (normal, then tentative, then unknown) or not is actually irrelevant (to the end user at least - Michele agreed that it makes sense to process separately).
The recalibration function can be calculated on the "knowns". This can then be used to adjust the masses in the subsequent tentatives and/or unknowns - this must be possible because even the masses of the fail peaks can be adjusted with the calibration. We don't want to recalculate the recalibration on tentatives (and we can't recalculate it on unknowns, like you said).
But, I still like the idea that one has to define explicitly whether to include tentatives or unknowns. The default workflow should certainly stay with the knowns (so the normal user gets an error if smiles are missing, for example).
From the user point of view, it's much easier if it's all bunched together in one compound list, one file list (the files are organised by machine runs and e.g. for us are mixed knowns, tentatives and unknowns) and I get one infolist and failpeak list in the end to check, not 3. Of course for the unknown we'd have to have no fail peaks... right? Because they'd all fail. Does this make sense?

**Comment by ermueller** [comment]
> I still like the idea that one has to define explicitly whether to include tentatives or unknowns.

If we're defining explicitely:

You could do the level numbers thing or you could write "tentative" or "unknown" in the last column of a compound list. This has the advantage that in a level 3 run where SMILES is present BUT tentative a COMMENT could be auto-added that says "Tentative spectrum".

Minor detail: If it's not the level numbers but "known", "tentative" or "unknown" (easier to understand for the user), what should the column name be?

Writing an auto-detect would also be no big deal, really, but then we couldn't do the level 3 autocomment since we wouldn't know whether the SMILES is correct or just tentative.

> Whether they are actually processed one after the other (normal, then tentative, then unknown) or not is actually irrelevant (to the end user at least - Michele agreed that it makes sense to process separately).

It definitely makes sense, but RMassBank is not structured to do several runs. Making the 3 runs separate is a solution that would work, but it'd probably take a bit longer and in the end would be just as complicated to program. I'd rather process on a compound by compound basis, since it does that already - I already changed quite a few internal functions to be able to discern between known, tentative and unknown, so I might as well walk the whole nine yards and do that for the rest as well.

I can undo the added "retrieval" parameter to msmsWorkflow, mbWorkflow and msmsRead and everything works internally. Then the only change for the user would be in the compound list.

> The recalibration function can be calculated on the "knowns".

Ahhhh, that makes sense. But, tentative with SMILES *has* to be explicitely stated then. Can't get around that?

**Comment by tsufz** [comment]
I also like the idea to tag the entries with the known, tentative and 
unknown tags. Might is possible to use both? Levels and tags? The levels 
are more advanced, but the tags are a 'quick and dirty' thing. Don't 
overload the users with requirements.

I agree with Erik to process in parallel if possible. I dunno what are 
the performance differences of separate and parallel processing. But 
from users view it is more reasonable to follow only one loop in the 
frontend. In the backend a stacked processing might be reasonable.

**Comment by schymane** [comment]
I already have the levels in my compound list and also already that level 3 case you mention :).
I see two possibilities: you leave the user free to type what they want, then the records are a mess because it's just text.
Or we specify exact words and levels and build the real comment behind the scenes, so the user could enter either 3 or tentative, for example. We'd have confirmed or reference standard (level 1), probable library (2a), probable diagnostic (2b), tentative (3), formula (4) or unknown (5). What do you think? 
The wording for the exact comment is a bit longer...and my problem is that I sub-categorised level 3 sometimes too. 
3a structure, 3b isomer, 3c substance class, 3d best match? But this is not written anywhere. 3a only if smiles are given.
________________________________________

**Comment by uchem-massbank** [comment]
Column name 'Confidence'?
________________________________________

**Comment by meowcat** [comment]
Do you think I can pull this branch to the main repo already?

**Comment by schymane** [comment]
@meowcat I suspect it will change but if it makes your life easier for now, I vote yes.
@ermueller I just emailed you a "translation table" csv for level, keyword and resulting comment, hope you get it but let me know if your email is still playing up...

**Comment by schymane** [comment]
w <- msmsRead(w, filetable="E:/UchemSpectra/Processed/AndreaTPs_ErikRMB/AndreaTPs_FileList_2014_10_tentatives.csv", mode=RMB_mode, readMethod = "mzR",retrieval="tentative")
> lapply(w@spectra,function(s) s@found)
$CYP1.mzML
[1] TRUE
$CYP1.mzML
[1] TRUE
.... 

Yay :)

**Comment by ermueller** [comment]
Thanks for the csv! I can definitely work with that.

@meowcat @uchem-massbank 
Please don't pull yet, I'll rework some stuff. Don't want "nonfinal" things in the main repo. It'll probably be finished next week if I don't hit any major roadblocks. :)

**Comment by uchem-massbank** [comment]
No probs! I’ve got my first set of tentative records using your workflow (much better calibration than for the knowns, because I had too few “knowns” => this really adds to the argument to do all together) and I’m about to try the first unknown. Pretty happy so far ;)

**Comment by ermueller** [comment]
Wait, but @schymane said:

> We don't want to recalculate the recalibration on tentatives (and we can't recalculate it on unknowns, like you said).

So, now I'm kinda confused at what exactly to do with tentatives in recalibration. You mean level 4 (formula only), correct? I mean, there definitely is the possibility that someone knows the formula for sure, but just not the structure, so why not use it in recal? On the other hand, if the formula turns out to be wrong that could be mighty bad for recal.

Introduce 4a and 4b? 4a "Definitive molecular formula", 4b "Likely molecular formula"?

**Comment by uchem-massbank** [comment]
My email account is playing tricks on my identity, sorry – half the time it answers as uchem-mb ;)
So – the rule with the recalibration is as long as we have a formula (or better still smiles) we can recalibrate. Irrespective of whether the structure is tentative or not. So, if it’s not Level 5, let’s use it in the recalibration. We really need this.
No to the multiple level 4s – level 4 is defined as an *unequivocal* molecular formula, if you are not sure what the formula is, it’s Level 5.

**Comment by schymane** [comment]
@ermueller RE: your email:

w <- msmsRead(w,mode="pH",files=w@files,readMethod="mzR",cpdids=c(1,2,3),retrieval="tentative")
w <- msmsWorkflow(w, mode="pH", steps=2:8,archivename="TEST", retrieval="tentative")
mb <- newMbWorkspace(w)
mb <- mbWorkflow(mb, 1:2, retrieval="tentative")
mb <- loadInfolist(mb,"./infolist.csv")
mb <- mbWorkflow(mb, 3:8, retrieval="tentative")

This seems to actually work fine without retrieval=”tentative” after the msmsRead stage. Which was nice ;)

**Comment by schymane** [comment]
@ermueller @meowcat I solved the error in the unknown file so you can ignore the email from yesterday - in this case retrieval="unknown" is essential at all steps, unlike retrieval="tentative" which I only needed for msmsRead. So now I have the records of the unknown too ;) and the recalibration code in the vignette works. Look forward to the next version!

**Comment by meowcat** [comment]
From looking at the code today when merging, it seemed to me that actually, the way the "standard" / "tentative" / "unknown" arguments are used is a bit redundant with the actual specifications in the compoundlist - the argument would not actually be needed when processing, since the entry in the compoundlist already gives it away whether a compound is a known, tentative or unknown identification... Just a thought...

**Comment by uchem-massbank** [comment]
I had assumed that this would be read from the compound list in the future version rather than defined explicitly, this would also make much more sense to me. But I guess Erik knows best what he is doing :-)

**Comment by ermueller** [comment]
Yes, I would remove these arguments from the functions. :) It wouldn't make any sense to doubly specify that stuff.

**Comment by ermueller** [comment]
You all got a mail. It's finished! - I hope.

I'm sorry, @meowcat, I just realized you weren't a recipient, because I forgot to add you. I just hit "reply all" and wrote in there on the big mail back-and-forth from yesterday. The new stuff is in my repo as well.

**Comment by schymane** [comment]
Great :-) Can’t wait to try it out. Just one question:
- Molfiles for formula and unknown result in empty files
Empty files, or no files? I wonder if empty files will create issues with MassBank upload…

**Comment by ermueller** [comment]
Empty files. I can change that, too?

**Comment by schymane** [comment]
Yes, would be much better to have no files at all and no entry in list.tsv…

**Comment by ermueller** [comment]
Did that, it's in my repo now with no molfiles at all and no entry in list.tsv

**Comment by schymane** [comment]
Cool! Btw, do you know how to build from github including the vignette?

**Comment by tsufz** [comment]
library(devtools)
install_github("ermueller/RMassBank")

**Comment by schymane** [comment]
..and if I do that it appears the default is not to include the vignette…hence my question…

**Comment by tsufz** [comment]
Maybe install_github("ermueller/RMassBank", build_vignettes = TRUE)


build_vignettes = F is default setting:

http://stackoverflow.com/questions/25821473/building-r-package-from-github-how-to-disable-building-vignettes

**Comment by schymane** [comment]
* creating vignettes ...Warning: running command '"C:/PROGRA~1/R/R-32~1.0/bin/x64/Rscript" --vanilla --default-packages= -e "tools::buildVignettes(dir = '.', tangle = TRUE)"' had status 1
ERROR

[omitted lots of stuff in the middle here]

Error: processing vignette 'RMassBankXCMS.Rnw' failed with diagnostics:

 chunk 9

Error : .onLoad failed in loadNamespace() for 'CAMERA', details:

  call: .setCAMERAOptions(pkgname)

  error: 4 arguments passed to .Internal(nchar) which requires 3

Execution halted

Error: Command failed (1)

**Comment by tsufz** [comment]
Of course, you need a basic TEX on your computer (if windows) to compile 
the PDF. You may use TEXLive:

http://www.tug.org/texlive/acquire-netinstall.html

Use custom install and select the basic scheme (plain+LATEX)

Good luck!

**Comment by ermueller** [comment]
Is that a TEX problem? Looks to me like a problem with CAMERA...

**Comment by schymane** [comment]
I thought the same ... the "normal" vignette compiles first and didn't give an error, which kinda rules out tex as the issue. Plus I have it anyway.
Can you try on your system? I had issues with rtools and packages in the lead up to that error as well, but seemed to be fixed.
The package installed OK without vignette.

**Comment by ermueller** [comment]
It works just fine when I do it. Huh, that's strange... You do have the newest CAMERA and xcms, right? That's the only thing I can think of.

**Comment by schymane** [comment]
well, since I had issues with a number of other packages prior to that it could well be an out-of-date CAMERA. Can try tomorrow and will let you know.

**Comment by schymane** [comment]
The problem is certainly CAMERA but I can’t fix it whatever I try.

Error: processing vignette 'RMassBankXCMS.Rnw' failed with diagnostics:
chunk 9
Error : .onLoad failed in loadNamespace() for 'CAMERA', details:
  call: .setCAMERAOptions(pkgname)
  error: 4 arguments passed to .Internal(nchar) which requires 3
Execution halted
Error: Command failed (1)



> library(CAMERA)

[…]

Error : .onLoad failed in loadNamespace() for 'CAMERA', details:

  call: .setCAMERAOptions(pkgname)

  error: 4 arguments passed to .Internal(nchar) which requires 3



> source("https://bioconductor.org/biocLite.R")

Bioconductor version 3.2 (BiocInstaller 1.20.1), ?biocLite for help

> biocLite("CAMERA")

Using Bioconductor 3.2 (BiocInstaller 1.20.1), R 3.2.0 (2015-04-16).

Installing package(s) ‘CAMERA’

trying URL 'https://bioconductor.org/packages/3.2/bioc/bin/windows/contrib/3.2/CAMERA_1.26.0.zip'

Content type 'application/zip' length 3040658 bytes (2.9 MB)

downloaded 2.9 MB



package ‘CAMERA’ successfully unpacked and MD5 sums checked

Warning: cannot remove prior installation of package ‘CAMERA’



The downloaded binary packages are in

        C:\Users\schymaem\AppData\Local\Temp\RtmpMl9VOB\downloaded_packages

Old packages: 'MASS', 'Matrix', 'mgcv', 'nlme'

Update all/some/none? [a/s/n]:

n [already updated them many times and it’s not working]

> library(CAMERA)

Error in library(CAMERA) : there is no package called ‘CAMERA’


> library("CAMERA", lib.loc="~/R/win-library/3.2")
Error in library("CAMERA", lib.loc = "~/R/win-library/3.2") :
  there is no package called ‘CAMERA’
> install.packages("C:/DATA/R/CAMERA_1.26.0.tar.gz", repos=NULL, type="source")
Installing package into ‘C:/DATA/R/R/win-library/3.2’
(as ‘lib’ is unspecified)
* installing *source* package 'CAMERA' ...
** libs

*** arch - i386
gcc -m32 -I"C:/PROGRA~1/R/R-32~1.0/include" -DNDEBUG     -I"d:/RCompile/r-compiling/local/local320/include"     -O3 -Wall  -std=gnu99 -mtune=core2 -c fastMatch.c -o fastMatch.o
gcc -m32 -shared -s -static-libgcc -o CAMERA.dll tmp.def fastMatch.o -Ld:/RCompile/r-compiling/local/local320/lib/i386 -Ld:/RCompile/r-compiling/local/local320/lib -LC:/PROGRA~1/R/R-32~1.0/bin/i386 -lR
installing to C:/DATA/R/R/win-library/3.2/CAMERA/libs/i386

*** arch - x64
gcc -m64 -I"C:/PROGRA~1/R/R-32~1.0/include" -DNDEBUG     -I"d:/RCompile/r-compiling/local/local320/include"     -O2 -Wall  -std=gnu99 -mtune=core2 -c fastMatch.c -o fastMatch.o
gcc -m64 -shared -s -static-libgcc -o CAMERA.dll tmp.def fastMatch.o -Ld:/RCompile/r-compiling/local/local320/lib/x64 -Ld:/RCompile/r-compiling/local/local320/lib -LC:/PROGRA~1/R/R-32~1.0/bin/x64 -lR
installing to C:/DATA/R/R/win-library/3.2/CAMERA/libs/x64
Warning in file.copy(files, dest, overwrite = TRUE) :
  problem copying .\CAMERA.dll to C:\DATA\R\R\win-library\3.2\CAMERA\libs\x64\CAMERA.dll: Permission denied
** R
** data
** inst
** byte-compile and prepare package for lazy loading
Warning: package 'Biobase' was built under R version 3.2.2
Warning: package 'BiocGenerics' was built under R version 3.2.2
Warning: package 'xcms' was built under R version 3.2.2
Warning: package 'mzR' was built under R version 3.2.2
Warning: package 'Rcpp' was built under R version 3.2.3
Warning in fun(libname, pkgname) :
  mzR has been built against a different Rcpp version (0.12.1)
than is installed on your system (0.12.3). This might lead to errors
when loading mzR. If you encounter such issues, please send a report,
including the output of sessionInfo() to the Bioc support forum at
https://support.bioconductor.org/. For details see also
https://github.com/sneumann/mzR/wiki/mzR-Rcpp-compiler-linker-issue.
Warning: package 'ProtGenerics' was built under R version 3.2.2
Warning: package 'igraph' was built under R version 3.2.2
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
*** arch - i386
Warning: package 'Biobase' was built under R version 3.2.2
Warning: package 'BiocGenerics' was built under R version 3.2.2
Warning: package 'xcms' was built under R version 3.2.2
Warning: package 'mzR' was built under R version 3.2.2
Warning: package 'Rcpp' was built under R version 3.2.3
Warning in fun(libname, pkgname) :
  mzR has been built against a different Rcpp version (0.12.1)
than is installed on your system (0.12.3). This might lead to errors
when loading mzR. If you encounter such issues, please send a report,
including the output of sessionInfo() to the Bioc support forum at
https://support.bioconductor.org/. For details see also
https://github.com/sneumann/mzR/wiki/mzR-Rcpp-compiler-linker-issue.
Warning: package 'ProtGenerics' was built under R version 3.2.2
Warning: package 'igraph' was built under R version 3.2.2
*** arch - x64
Warning: package 'Biobase' was built under R version 3.2.2
Warning: package 'BiocGenerics' was built under R version 3.2.2
Warning: package 'xcms' was built under R version 3.2.2
Warning: package 'mzR' was built under R version 3.2.2
Warning: package 'Rcpp' was built under R version 3.2.3
Warning in fun(libname, pkgname) :
  mzR has been built against a different Rcpp version (0.12.1)
than is installed on your system (0.12.3). This might lead to errors
when loading mzR. If you encounter such issues, please send a report,
including the output of sessionInfo() to the Bioc support forum at
https://support.bioconductor.org/. For details see also
https://github.com/sneumann/mzR/wiki/mzR-Rcpp-compiler-linker-issue.
Warning: package 'ProtGenerics' was built under R version 3.2.2
Warning: package 'igraph' was built under R version 3.2.2
* DONE (CAMERA)
> library(CAMERA)
Error in get0(hookname, envir = env, inherits = FALSE) :
  lazy-load database 'C:/DATA/R/R/win-library/3.2/CAMERA/R/CAMERA.rdb' is corrupt
In addition: Warning message:
In get0(hookname, envir = env, inherits = FALSE) :
  internal error -3 in R_decompress1
Error: package or namespace load failed for ‘CAMERA’




R version 3.2.0 (2015-04-16)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 8 x64 (build 9200)

locale:
[1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252    LC_MONETARY=English_Australia.1252
[4] LC_NUMERIC=C                       LC_TIME=English_Australia.1252

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] igraph_1.0.1         xcms_1.46.0          ProtGenerics_1.2.1   mzR_2.4.0            Rcpp_0.12.3
[6] Biobase_2.30.0       BiocGenerics_0.16.1  devtools_1.9.1       BiocInstaller_1.20.1

loaded via a namespace (and not attached):
[1] RColorBrewer_1.1-2  plyr_1.8.3          tools_3.2.0         rpart_4.1-10        digest_0.6.9
 [6] memoise_0.2.1       gtable_0.1.2        lattice_0.20-33     graph_1.48.0        curl_0.9.4
[11] gridExtra_2.0.0     httr_1.0.0          stringr_1.0.0       cluster_2.0.3       nnet_7.3-11
[16] stats4_3.2.0        grid_3.2.0          R6_2.1.1            RBGL_1.46.0         survival_2.38-3
[21] foreign_0.8-66      latticeExtra_0.6-26 Formula_1.2-1       ggplot2_2.0.0       magrittr_1.5
[26] Hmisc_3.17-1        scales_0.3.0        codetools_0.2-14    splines_3.2.0       colorspace_1.2-6
[31] stringi_1.0-1       acepack_1.3-3.3     munsell_0.4.2


>




So, does anyone have any brilliant ideas?

**Comment by tsufz** [comment]
I had problems with the Rcpp version. You may reinstall the 0.12.1 version:

packageurl <- 
"http://cran.r-project.org/src/contrib/Archive/Rcpp/Rcpp_0.12.1.tar.gz"
install.packages(packageurl, repos=NULL, type="source")

**Comment by ermueller** [comment]
Oh god, R installing problems.

> Warning: package 'Biobase' was built under R version 3.2.2 
> Warning: package 'BiocGenerics' was built under R version 3.2.2 
> Warning: package 'xcms' was built under R version 3.2.2 
> Warning: package 'mzR' was built under R version 3.2.2 
> Warning: package 'Rcpp' was built under R version 3.2.3

Some version problem? I dunno why, you're definitely installing CAMERA for R 3.2

remove 
'C:/DATA/R/R/win-library/3.2/CAMERA/'

just completely delete that whole folder, since some file is corrupt.

Another thing:

> Warning in file.copy(files, dest, overwrite = TRUE) : problem copying .\CAMERA.dll to C:\DATA\R\R\win-library\3.2\CAMERA\libs\x64\CAMERA.dll: Permission denied

You don't have complete access to that folder? I think an installation will fail as long as you do that. My oh my, this mess gives me flashbacks. I had that, too, when I got my new laptop. But that is the rigth library path, right?

**Comment by meowcat** [comment]
That CAMERA error is what I had when I tried to install (github) RMassBank-refactored on R 3.2. Which prompted me to upgrade completely to R 3.3 and Bioconductor-devel. I don't think there is another solution, we can't backport everything to 3.2...

**Comment by tsufz** [comment]
Hi, the problem wit Rcpp and mzR also persists. mzR 2.40 is compiled 
with Rcpp 0.12.1, but 0.12.3 is required for RMassBank. Would be nice to 
get a mzR version compiled with Rcpp 0.12.3 during installation to get 
rid of the problems.

**Comment by schymane** [comment]
So, I have permission to that folder (in principle at least...) and CAMERA is not in the library in the folder I do not have permission to access (don't get me started on our admin rights policy...)
So, I deleted the CAMERA folder and did:

> source("https://bioconductor.org/biocLite.R")
Bioconductor version 3.2 (BiocInstaller 1.20.1), ?biocLite for help
> biocLite("CAMERA")
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.2 (BiocInstaller 1.20.1), R 3.2.0 (2015-04-16).
Installing package(s) ‘CAMERA’
trying URL 'https://bioconductor.org/packages/3.2/bioc/bin/windows/contrib/3.2/CAMERA_1.26.0.zip'
Content type 'application/zip' length 3040658 bytes (2.9 MB)
downloaded 2.9 MB

package ‘CAMERA’ successfully unpacked and MD5 sums checked

The downloaded binary packages are in
        C:\Users\schymaem\AppData\Local\Temp\RtmpQDeIdJ\downloaded_packages
Old packages: 'MASS', 'Matrix', 'mgcv', 'nlme'
Update all/some/none? [a/s/n]:
n
> library(CAMERA)
Loading required package: Biobase
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, parApply,
    parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from ‘package:stats’:

    IQR, mad, xtabs

The following objects are masked from ‘package:base’:

    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames, do.call, duplicated, eval, evalq,
    Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rownames, sapply, setdiff,
    sort, table, tapply, union, unique, unlist, unsplit

Welcome to Bioconductor

    Vignettes contain introductory material; view with 'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.

Loading required package: xcms
Loading required package: mzR
Loading required package: Rcpp
Loading required package: ProtGenerics

Attaching package: ‘xcms’

The following objects are masked from ‘package:Biobase’:

    phenoData, phenoData<-

Loading required package: igraph

Attaching package: ‘igraph’

The following object is masked from ‘package:xcms’:

    groups

The following objects are masked from ‘package:BiocGenerics’:

    normalize, union

The following objects are masked from ‘package:stats’:

    decompose, spectrum

The following object is masked from ‘package:base’:

    union

Error : .onLoad failed in loadNamespace() for 'CAMERA', details:
  call: .setCAMERAOptions(pkgname)
  error: 4 arguments passed to .Internal(nchar) which requires 3
In addition: Warning messages:
1: package ‘CAMERA’ was built under R version 3.2.2
2: package ‘Biobase’ was built under R version 3.2.2
3: package ‘BiocGenerics’ was built under R version 3.2.2
4: package ‘xcms’ was built under R version 3.2.2
5: package ‘mzR’ was built under R version 3.2.2
6: package ‘Rcpp’ was built under R version 3.2.3
7: In fun(libname, pkgname) :
  mzR has been built against a different Rcpp version (0.12.1)
than is installed on your system (0.12.3). This might lead to errors
when loading mzR. If you encounter such issues, please send a report,
including the output of sessionInfo() to the Bioc support forum at
https://support.bioconductor.org/. For details see also
https://github.com/sneumann/mzR/wiki/mzR-Rcpp-compiler-linker-issue.
8: package ‘ProtGenerics’ was built under R version 3.2.2
9: package ‘igraph’ was built under R version 3.2.2
Error: package or namespace load failed for ‘CAMERA’




So then when I try to install the source, it doesn't work because of permission denied again. When I try to delete the folder again in windows  - also not because R blocked it! haha - that was the problem. New session of R then

install.packages("C:/DATA/R/CAMERA_1.26.0.tar.gz", repos=NULL, lib="C:/DATA/R/R/win-library/3.2", type="source")

* DONE (CAMERA)

and library(CAMERA) seems to work now. Hooray. So the fail to install locked access to the folder because it was still "in use".


> library(devtools)
Warning message:
package ‘devtools’ was built under R version 3.2.2
> install_github("ermueller/RMassBank", build_vignettes = TRUE)

and guess what?

* DONE (RMassBank)
> library(RMassBank)
Warning messages:
1: replacing previous import by ‘ggplot2::Position’ when loading ‘MSnbase’
2: replacing previous import by ‘ggplot2::unit’ when loading ‘MSnbase’
3: replacing previous import by ‘ggplot2::arrow’ when loading ‘MSnbase’
4: In .recacheSubclasses(def@className, def, doSubclasses, env) :
  undefined subclass "externalRefMethod" of class "expressionORfunction"; definition not updated
5: In .recacheSubclasses(def@className, def, doSubclasses, env) :
  undefined subclass "externalRefMethod" of class "functionORNULL"; definition not updated
> vignette("RMassBank")
 vignette("RMassBankXCMS")
All works. Finally. So manual delete of CAMERA dir and hand-install.


________________________________

**Comment by schymane** [comment]
Sounds to me like @sneumann needs to look at why CAMERA doesn't work properly for 3.2...

**Comment by schymane** [comment]
The problem or the warning?
The warning is annoying, but not fatal, I was not aware of it causing problems/errors?
________________________________

**Comment by ermueller** [comment]
Yeah, the warning is just kind of there for me. Never had any actual problems or errors.

**Comment by meowcat** [comment]
Hi Erik,

really cool. I am just running a few tests (not everything perfect yet but I want to examine the user error component first :) ) 

One suggestion, what about changing findMz:
```R
findMz <- function(cpdID, mode="pH", ppm=10, deltaMz=0)
{
	retrieval <- findLevel(cpdID, TRUE)
    # In case of unknown: m/z is in table
```
etc? so you could take out the retrieval argument from many more places.

I merged everything into master here already.

**Comment by meowcat** [comment]
**wrt using neutral or charged masses in the compoundlist mz column,** this is a bit of a question of definition. 
Technically, findMz should return the mass according to the "mode" set ("pH", "mH" etc) so using the neutral mass in the compoundlist would be more consistent. However, I don't know what @schymane thinks about this, since "in our heads" we typically think of the m/z and not of the uncharged mass; also there is no guarantee for level 5 (which means no certain formula) that m/z - massH + electron is actually the neutral M. Opinions?

**Comment by meowcat** [comment]
Also, findFormula() could do findLevel by itself without needing the argument.

**Comment by sneumann** [comment]
I can confirm the installation issue on Windows with biocLite(CAMERA).
The issue during installation with biocLite() seems to be specific to R-3.2.0,
as mentioned here https://stackoverflow.com/questions/30955508/r-cannot-load-package-forecast-due-to-namespace-error/30956549#30956549,
and with a technical explanation in this comment>
http://stackoverflow.com/questions/32912268/error-in-r-data-table-v1-9-6-function-fread/32959306#32959306

Bottomline is that Windows binary ZIP packages built recently on R-3.2.>=1 
do not work on R-3.2.0, that explains why your build from source succeeded. 

````
install.packages("C:/DATA/R/CAMERA_1.26.0.tar.gz", repos=NULL, lib="C:/DATA/R/R/win-library/3.2", type="source")
...
* DONE (CAMERA)
````
Glad the install worked in the end on R-3.2.

The Rcpp is only a warning, and only causes issues if the Rcpp people 
changed their API and/or internals. They do announce that upfront, 
but I haven't heard anything that should be of concern lately. 
Yours, Steffen

**Comment by schymane** [comment]
Yes, good point. The Level 5 mass is indeed the charged one (“the exact mass of interest”) and therefore would not need the mode correction – which is counter-intuitive to the mode setting I agree. But if we change to the neutral mass, users will have to “guess” the adduct state to input a neutral mass that will mean that RMassBank can calculate back to the exact mass they were interested in in the first place and this could get messy… I’d see this as more of a risk.
I’d vote to stick to the level 5 being the charged mass (“exact mass of interest”) and it should be clearly stated somewhere that level 5 masses are NOT adjusted according to the mode settings?

**Comment by meowcat** [comment]
@everyone: I branched off all the issues with installation that are unrelated to this question so this stays a bit more focused... #149

**Comment by meowcat** [comment]
after a bit of discussion with @schymane in the office:

on one point you are right about the mass of interest often being the m/z, on the other hand this depends on the workflow (are you using a nontarget list, are you recording spectra of tentatively identified compounds etc.) For example,  I have an MS/MS method that records 9 positive and 9 negative MS2 for every compound in the file. Here it would be much easier to process both with mode "pH" then "mH" starting from the neutral mass.

The first solution we were talking about would be to add one more column to the compound list that specifies *what* is in the mz column, or a parameter to loadList which specifies how to interpret the mz column. However I have another idea now - can we just check what the column is called? so if the column is called m/z or mz it is a charged mass, and if the column is called "mass" it is the neutral mass. Would this be confusing?

**Comment by schymane** [comment]
What about mz and neutral_mass – then there is no room for doubt?

**Comment by meowcat** [comment]
fixed a bug in mbworkflow step 6 (in this repo)

**Comment by ermueller** [comment]
Yeah, mz or "m/z" and neutral_mass sound great.

Currently I'm calculating the mass with findMass by doing this:
````
if(mode == "pH") {
            mass <- 1.00784
            mode.charge <- 1
        } else if(mode == "pNa") {
            mass <- 22.989769
            mode.charge <- 1
        } else if(mode == "pM") {
            mass <- 0
            mode.charge <- 1
        } else if(mode == "mM") {
            mass <- 0
            mode.charge <- -1
        } else if(mode == "mH") {
            mass <- -1.00784
            mode.charge <- -1
        } else if(mode == "mFA") {
            mass <- 59.0440
            mode.charge <- -1
        } else if(mode == "pNH4") {
            mass <- 18.03846
            mode.charge <- 1
        }
````
and then calling

````
findMz(cpdID_or_smiles, mode=mode, retrieval=retrieval)$mzCenter - mass + mode.charge * .emass
````

or is that bad?

Also @meowcat
yes, I will remove the retrieval parameter for these functions, I dunno what I was thinking. Didn't see the forest for the trees there.

**Comment by meowcat** [comment]
> Currently I'm calculating the mass with findMass by doing this:

One could just try inside findMz:
```R
    if(retrieval == "unknown"){
        mz <- .listEnvEnv$listEnv$compoundList[which(.listEnvEnv$listEnv$compoundList$ID == cpdID),.listEnvEnv$listEnv$mzCol]
        if(neutralblabla)
        {
                shift <- findMz.formula("", mode)
                mz <- mz + shift
        }
        # etc
        }
```

**Comment by ermueller** [comment]
That doesn't work for "mM"(Error), or "mH"(Just returns 1.0000X when it should return -1.0000X)

**Comment by ermueller** [comment]
Oh and pM either. But I could just handle these cases in an "if".

**Comment by meowcat** [comment]
Hm but this means my function is working incorrectly. I will fix this :)

**Comment by meowcat** [comment]
Try the fix in master branch d05725f

**Comment by meowcat** [comment]
Digging up this old one. I am using tentative processing routinely. But we still don't have the "neutral_mass" option, right? Or is it in a branch I don't have? I would need it urgently...

**Comment by schymane** [comment]
@meowcat is now building in the option to treat the mass as either neutral or charged … we just talked about it

**Comment by meowcat** [comment]
As @schymane said, I am fixing this (currently in the S4power branch) with an option in the settings file, which is also a parameter to the findMz function:
```
unknownMass: charged # (or absence of the option): behaviour as up to now (the mass is m/z)
unknownMass: neutral # the mass is treated as neutral mass and +H-e or -H+e are applied accordingly.
```

**Comment by meowcat** [comment]
@schymane @tsufz Vignette for processing of unknowns?

**Comment by schymane** [comment]
Sounds like a good idea ... do we not have it in the nonstandard vignette?

**Comment by tsufz** [comment]
fixed in PR https://github.com/MassBank/RMassBank/pull/231
